feat: document using compressed files (#656)
opensafely/documentation: docs/actions-scripts.md

* **Do not write code that requires an internet connection to run.**
Any research objects (datasets, libraries, etc.) that are retrieved via the internet should be imported to the repo locally first.
If this is not possible (for instance if the object size is too large to be transferred via GitHub) then get in touch.

* **Avoid code that consumes a lot of time or memory.** The server is not an infinite resource. We can advise on code optimisation if run-times become problematic.  A good strategy is to split your processing into separate project pipeline actions; the job runner can then choose to run them in parallel if sufficient resources are available.

* **Write code that runs in the OpenSAFELY platform.**
Code will be run within a Linux-based Docker environment. In practice, this just means ensuring you use forward-slashes `/` for directories.

* **Structure your code into discrete chunks, both within scripts, and by splitting into different pipeline actions.**
This helps with:
	* readability
	* bug-finding
	* parallelisation via the project pipeline

## Reading and Writing Outputs

> Scripted actions can read and write output files that are persisted in the workspace. These generally fall into two categories: large files of `highly_sensitive` data to feed into other actions, or smaller `moderately_sensitive` outputs intended for review and release.

Reviewer: s/persisted/saved? Or maybe expand as to what persisted means (which may clarify to the reader why overly large files are an issue)?

Author: Good call.

### Large `highly_sensitive` Output Files

It is important that the right files formats are used for large data files. The wrong formats can waste disk space, execution time, and server memory. The specific formats used vary with language ecosystem, but they should always be compressed.

Note: the `cohortextractor` command produces `csv.gz` outputs, which is the current recommended output to be consumed by all supported languages.

=== "Python"

    ```python
    # read compressed csv output from cohortextractor
    pd.read_csv("output/input.csv.gz)

    # read feather file, decompressed automatically
    pd.read_feather("output/input.feather")

    # write compressed feather file

>     df.to_feather("output/model.feather", compression="zstd")

Reviewer: isn't [lz4 compression enabled by default](https://arrow.apache.org/docs/python/feather.html#using-compression), or is it inadequate in some way?

Author: It is, but it isn't as good as zstd in terms of compression ratio, whilst being nearly as fast. And we have a lot data. If I could change the default I would.

    ```

=== "R"

    ```r
    # read compressed csv output from cohortextractor
    df <- readr::read_csv("output/input.csv.gz")

>     # read a feather file, decompressed automatically

Reviewer: can cohortextractor also produce feather files? My reading of this implies that it cannot

Reviewer: or is it just for inter-scripted action data (as per .dta.gz in stata)?

Author: It can, but we are unsure that feather is ready to be the default just yet.

The intention here was to read feather files in subsequent actions. 

Perhaps I can add switch the order of feather read/write examples with a comment

Author: I have switched the order and added a comment for clarity

    df <- arrow::read_feather("output/input.feather")

    # write a compressed feather file
    arrow::write_feather(df, "output/model.feather", compression = "zstd")
    ```

=== "Stata"

    ```stata
    // stata cannot handle compressed csv files directly, so unzip first to a plain csv file
    // the unzipped file will be discarded when the action finishes.
    !gunzip output/input.csv.gz
    // now import the uncompressed csv using delimited
    import delimited using output/input.csv

    // load a compressed .dta.gz file
    gzload output/input.dta.gz

    // save in compressed dta.gz format
    gzsave output/model.dta.gz
    ```

### Smaller `moderately_senstive` Output Files

> These outputs are marked as `moderately_sensitive` in your `project.yaml`, and are available to view with [level 4 access](level-4-server.md). They can consist of aggregate summary data, images, or log files for debugging action code.

Reviewer: "Level 4" is used elsewhere - do we standardise on casing?

Due to the fact that level 4 files need to be reviewed, there are various restrictions placed on sizes and formats of files that can be released

Firstly, only certain formats are supported, in order that the reviewers can properly read the outputs.

| Type | Formats |
| --- | --- |
| Text |  `.txt`, `.log`, `.md` |
| Data | `.csv`, `.tsv`, `.json` |
| Images | `.png`, `.jpeg`, `.svgz` |
| Reports | `.html`, `.pdf` |

> Additionally, there is a maximum file size of 32Mb, in order to a) limit the amount of data that can be accessed via level 4, and b) to ensure review of an output is tractable for reviewers.

Reviewer: 32Mb or 32MB?

Author: MB, fixed

* **Do not write code that requires an internet connection to run.**
Any research objects (datasets, libraries, etc.) that are retrieved via the internet should be imported to the repo locally first.
If this is not possible (for instance if the object size is too large to be transferred via GitHub) then get in touch.

* **Avoid code that consumes a lot of time or memory.** The server is not an infinite resource. We can advise on code optimisation if run-times become problematic.  A good strategy is to split your processing into separate project pipeline actions; the job runner can then choose to run them in parallel if sufficient resources are available.

* **Write code that runs in the OpenSAFELY platform.**
Code will be run within a Linux-based Docker environment. In practice, this just means ensuring you use forward-slashes `/` for directories.

* **Structure your code into discrete chunks, both within scripts, and by splitting into different pipeline actions.**
This helps with:
	* readability
	* bug-finding
	* parallelisation via the project pipeline

## Reading and Writing Outputs

> Scripted actions can read and write output files that are saved in the workspace. These generally fall into two categories: large files of `highly_sensitive` data to feed into other actions, or smaller `moderately_sensitive` outputs intended for review and release.

Reviewer 2: suggestion:

I tend to favour using lists in technical writing where possible. Feel free to decide whether you prefer this or ignore it.

```suggestion
Scripted actions can read and write output files that are saved in the workspace. These generally fall into two categories:
* large files of `highly_sensitive` data for use by other actions
* smaller `moderately_sensitive` outputs for review and release
```

> ### Large `highly_sensitive` Output Files

Reviewer 2: ```suggestion
### Large `highly_sensitive` output files
```

It is important that the right files formats are used for large data files. The wrong formats can waste disk space, execution time, and server memory. The specific formats used vary with language ecosystem, but they should always be compressed.

> Note: the `cohortextractor` command produces `csv.gz` outputs, which is the current recommended output to be consumed by all supported languages.

Reviewer 2: suggestion:

Move this to an admonition:

```suggestion
!!! note
    The `cohortextractor` command produces `csv.gz` outputs. `csv.gz` is the recommended output format.
```

I'd also be inclined to link to an explanation of:

> `csv.gz` is the recommended output format.

but we don't have a definitive section. It's part of the `docs/actions-cohortextractor.md` content, but not in its own separate section. This isn't too important though.

=== "Python"

    ```python
    # read compressed csv output from cohortextractor

>     pd.read_csv("output/input.csv.gz)

Reviewer 2: nit:

```suggestion
    pd.read_csv("output/input.csv.gz")
```

This is a good case for us testing all our code snippets end-to-end, as we tried out for Data Builder. It's very easy to make errors when editing source code within Markdown: the usual IDE/editor tooling that might flag may not work in this context.

    # write compressed feather file
    df.to_feather("output/model.feather", compression="zstd")

    # read feather file, decompressed automatically
    pd.read_feather("output/input.feather")
    ```

=== "R"

    ```r
    # read compressed csv output from cohortextractor
    df <- readr::read_csv("output/input.csv.gz")

    # write a compressed feather file
    arrow::write_feather(df, "output/model.feather", compression = "zstd")

    # read a feather file, decompressed automatically
    df <- arrow::read_feather("output/input.feather")
    ```

=== "Stata"

    ```stata
    // stata cannot handle compressed csv files directly, so unzip first to a plain csv file
    // the unzipped file will be discarded when the action finishes.
    !gunzip output/input.csv.gz
    // now import the uncompressed csv using delimited
    import delimited using output/input.csv

    // save in compressed dta.gz format
    gzsave output/model.dta.gz

    // load a compressed .dta.gz file
    gzload output/input.dta.gz

>     ```

Reviewer 2: thought: This is all relatively simple and looks plausible. I don't know R or Stata really, so there may be something I've missed.

> ### Smaller `moderately_senstive` Output Files

Reviewer 2: ```suggestion
### Smaller `moderately_sensitive` output files
```

> These outputs are marked as `moderately_sensitive` in your `project.yaml`, and are available to view with [Level 4 access](level-4-server.md). They can consist of aggregate summary data, images, or log files for debugging action code.

Reviewer 2: ```suggestion
These outputs are marked as `moderately_sensitive` in your `project.yaml`, and are available to view with [Level 4 access](level-4-server.md). Outputs can consist of:
* aggregate summary data
* images
* log files for debugging action code
```

Due to the fact that Level 4 files need to be reviewed, there are various restrictions placed on sizes and formats of files that can be released

> Firstly, only certain formats are supported, in order that the reviewers can properly read the outputs.

Reviewer 2: ```suggestion
#### File format restrictions

These are restricted so that reviewers can properly examine the outputs on the secure server.
```

| Type | Formats |
| --- | --- |
| Text |  `.txt`, `.log`, `.md` |
| Data | `.csv`, `.tsv`, `.json` |
| Images | `.png`, `.jpeg`, `.svgz` |
| Reports | `.html`, `.pdf` |

> Additionally, there is a maximum file size of 32MB, in order to a) limit the amount of data that can be accessed via Level 4, and b) to ensure review of an output is tractable for reviewers.

Reviewer 2: ```suggestion
#### File size restrictions

There is a maximum file size of 32 MB to:

* limit the amount of data that can be accessed via Level 4
* allow a thorough review of the outputs in a reasonable time
```

> 
> Additionally, there is a maximum file size of 32MB, in order to a) limit the amount of data that can be accessed via Level 4, and b) to ensure review of an output is tractable for reviewers.
> 
> 
> 

Reviewer 2: ```suggestion
#### File size restrictions

There is a maximum file size of 32 MB to:

* limit the amount of data that can be accessed via Level 4
* allow a thorough review of the outputs in a reasonable time

```

As Stata is a commercial product, a license key is needed to use it. 

> #### If you are a member of the OpenSAFELY GitHub organisation

Reviewer: inconsistent with https://github.com/opensafely/documentation/pull/656/files#r832064418

Author: I didn't intentially change this, will revert

* If you are using Windows, then the `opensafely` command line software will
automatically use the OpenSAFELY Stata license. 

> * If you are using MacOS:

Reviewer: Indent below?

Author: I didn't intentially change this, will revert

>    1. Download and install [GitHub's command-line tool (`gh`)](https://cli.github.com/)

Reviewer: these are rendering as bullets not numbered

Author: I didn't intentially change this, will revert

   2. Run `gh auth login --web`. Select the "HTTPS" option, and follow the instructions
  * The `opensafely` command line software will now automatically use the OpenSAFELY Stata license

#### All other external users

> If you are not a member of the `opensafely GitHub organisation, you must provide your own **Stata/MP** license. Unfortunately other Stata flavours are not yet supported; [let us know](how-to-get-help.md) if this is a problem.

Reviewer: "`opensafely" needs closing backtick

Reviewer: or [OpenSAFELY GitHub organisation](https://github.com/opensafely) ?

Author: I didn't intentially change this, will revert
